master/buildbot/reporters/base.py

```diff
@@ -87,17 +86,24 @@ def _does_generator_want_key(self, generator, key):
                 return True
         return False
 
+    def _get_chain_key_for_event(self, key, msg):
+        if key[0] in ["builds", "buildrequests"]:
+            return ("buildrequestid", msg["buildrequestid"])
+        return None
+
     @defer.inlineCallbacks
     def _got_event(self, key, msg):
-        pending_got_event_call = self._pending_got_event_call
-
-        # Mark this call as pending.
-        self._pending_got_event_call = d = defer.Deferred()
-
-        # Wait for previously pending call, if any, to ensure
-        # reports are sent out in the order events were queued.
-        if pending_got_event_call is not None:
-            yield pending_got_event_call
+        chain_key = self._get_chain_key_for_event(key, msg)
+        if chain_key is not None:
+            d = defer.Deferred()
+            pending_got_event_call = self._pending_got_event_calls.get(
+                chain_key
```

Comment:
I think this could fit in a single line. `pending_got_event_call` could be named `pending_call` to shorten the line as well.